File: 3988-create-grid-with-exactly-k-paths-i/3988-create-grid-with-exactly-k-paths-i.cpp
class Solution {
public:
    vector<string> createGrid(int m, int n, int k) {
        if (n == 1 || m == 1) {
            if (k != 1) {
                return {};
            }
            if (m == 1) {
                return { string(n, '.') };
            }
            if (n == 1) {
                return vector<string>(m, ".");
            }
        }
        
        if (k == 1) {
            vector<string> ans;
            ans.push_back(string(n, '.'));
            for (int i = 0; i < m - 1; ++i) {
                ans.push_back(string(n - 1, '#') + ".");
            }
            return ans;
        }
            
        else if (k == 2) {
            if (n >= 2 && m >= 2) {
                vector<string> ans;
                ans.push_back(string(n, '.'));
                ans.push_back(string(n - 2, '#') + "..");
                for (int i = 0; i < m - 2; ++i) {
                    ans.push_back(string(n - 1, '#') + ".");
                }
                return ans;
            }
            return {};
        }
            
        else if (k == 3) {
            if (n >= 3 && m >= 2) {
                vector<string> ans;
                ans.push_back(string(n, '.'));
                ans.push_back(string(n - 3, '#') + "...");
                for (int i = 0; i < m - 2; ++i) {
                    ans.push_back(string(n - 1, '#') + ".");
                }
                return ans;
            } else if (m >= 3 && n >= 2) {
                vector<string> ans;
                ans.push_back(string(n, '.'));
                ans.push_back(string(n - 2, '#') + "..");
                ans.push_back(string(n - 2, '#') + "..");
                for (int i = 0; i < m - 3; ++i) {
                    ans.push_back(string(n - 1, '#') + ".");
                }
                return ans;
            }
            return {};
        }
            
        else if (k == 4) {
            if (n >= 3 && m >= 3) {
                vector<string> ans;
                ans.push_back(string(n - 1, '.') + "#");
                ans.push_back(string(n - 3, '#') + "...");
                ans.push_back(string(n - 3, '#') + "#..");
                
                for (int i = 0; i < m - 3; ++i) {
                    ans.push_back(string(n - 1, '#') + ".");
                }
                return ans;
            }
            else if (m == 2 && n >= 4) {
                return {
                    string(4, '.') + string(n - 4, '#'),
                    string(n, '.')
                };
            }
            else if (n == 2 && m >= 4) {
                vector<string> ans;
                for (int i = 0; i < m; ++i) {
                    if (i < 4) {
                        ans.push_back("..");
                    } else {
                        ans.push_back("#.");
                    }
                }
                return ans;
            }
            return {};
        }
            
        return {};
    }
};

File: 3988-create-grid-with-exactly-k-paths-i/3988-create-grid-with-exactly-k-paths-i.py
class Solution:
    def createGrid(self, m: int, n: int, k: int) -> list[str]:
        if n == 1 or m == 1:
            if k != 1:
                return []
            if m == 1:
                return ['.' * n]
            if n == 1:
                return ['.']*m
        if k == 1:
            ans = ['.' * n]
            for _ in range(m - 1):
                ans.append('#' * (n - 1) + '.')
            return ans
        elif k == 2:
            if n >= 2 and m >= 2:
                ans = ['.' * n]
                ans.append('#' * (n - 2) + '..')
                for _ in range(m - 2):
                    ans.append('#' * (n - 1) + '.')
                return ans
            return []
        elif k == 3:
            if n >= 3 and m >= 2:
                ans = ['.' * n]
                ans.append('#' * (n - 3) + '...')
                for _ in range(m - 2):
                    ans.append('#' * (n - 1) + '.')
                return ans
            elif m >= 3 and n >= 2:
                ans = ['.' * n]
                ans.append('#' * (n - 2) + '..')
                ans.append('#' * (n - 2) + '..')
                for _ in range(m - 3):
                    ans.append('#' * (n - 1) + '.')
                return ans
            return []
        elif k == 4:
            if n >= 3 and m >= 3:
                ans = ['.' * (n - 1) + '#']
                ans.append('#' * (n - 3) + '...')
                ans.append('#' * (n - 3) + '#..')
                for _ in range(m - 3):
                    ans.append('#' * (n - 1) + '.')
                return ans
            elif m == 2 and n >= 4:
                return [
                    '.' * 4 + '#' * (n - 4),
                    '.' * n                  
                ]
            elif n == 2 and m >= 4:
                ans = []
                for i in range(m):
                    if i < 4:
                        ans.append('..')    
                    else:
                        ans.append('#.')     
                return ans
                
            return []
        return []